evdev/src/Evdev.hs

```diff
@@ -195,9 +208,15 @@ toCTimeVal t = LL.CTimeVal n (round $ f * 1_000_000)
 
 -- | Create a device from a valid path - usually /\/dev\/input\/eventX/ for some /X/.
 newDevice :: RawFilePath -> IO Device
-newDevice path = do
-    dev <- cErrCall "newDevice" path $ LL.newDevice path
-    return $ Device dev path
+newDevice path = newDeviceFromFd (Just path) =<< openFd path ReadOnly Nothing defaultFileFlags
+
+-- | General constructor for a Device.
+-- WARNING: The resulting `Device' will have the GC close your `Fd'.
+newDeviceFromFd :: Maybe RawFilePath -> Fd -> IO Device
+newDeviceFromFd path =
+  let path' = maybe "" id path
+  -- It is probably bad manners to flip a record type.
+  in fmap (flip Device path') . cErrCall "newDeviceFromFd" path' . LL.newDeviceFromFd
```

Comment:
I would prefeer not requiring a `Maybe RawFilePath`.  After researching a bit last night it appears that we could use `fcntl` with the `F_GETPATH` command to get `maybe a path` which might also be `not the only path` associated with a device.

That would really be good enough for me, but `System.Posix.File` does not supply the behavior in question.  I don't mind adding some more C to the project to get that functionality, but I figured I would get your opinion about it first ;)

Reply:
I'd be happy to remove that field (which means `Device` may as well just be a re-export of `LowLevel.Device`. Pretty sure `libevdev` exposes a function we could wrap as `Device -> IO RawFilePath` anyway.

Obviously that's a breaking change, but there'll be a few of those in the next release anyway.

Reply:
Cool.   I had the urge to do that when working on this, but I figured I would try to keep backwards compatibility.

Reply:
> Pretty sure libevdev exposes a function we could wrap as Device -> IO RawFilePath anyway.

Looks like that's [libevdev_get_phys](https://www.freedesktop.org/software/libevdev/doc/latest/group__bits.html#ga5df483b0b24d7b96ea1181808fab851d), which we already have. Worth checking that it seems to return the expected result.

Reply:
> Looks like that's libevdev_get_phys

Actually that's definitely not it. Looks like we will need to write some C. Hopefully should work as expected given that we have control of the FD.

Reply:
Oh, so `deviceName` wont get the path for us?

I didn't see your above comments until now.
In the [last commit](https://github.com/georgefst/evdev/pull/11/commits/7a089eb500d13233bae00537bb78b1d57d8b9b8c) I used `LL.deviceName`. I also left the field in, because I didn't want to lose the `Show Errno` instance.

I just checked and `LL.deviceName` doesn't give me the path I used to create the `Fd`, but it is still a useful name... :/

So, you think I should follow the `fnctl` with `F_GETPATH` route? I don't mind doing it, but I think it is worth noting that there is no guarantee that linux will give the path we are expecting for a given file descriptor.  They can have multiple paths associated with them, and deeper exploration reveals things can get even weirder.

I wouldn't mind just changing the field name to `deviceName` to be clear about it, but if you really want to be able to get the original path I will include the C code for a `fnctl` call.  The only use case I can really see for wanting the original path (or at least a valid path to the device) is if we want to try opening a device again after a failure (like waiting for a user to plug something in again after it is unplugged).  I guess it could also make better logs in some situations... ?

IDK, let me know what you think in response to my thoughts here, and then I'll move forward ;)

Reply:
I just checked (not in Evdev, but by calling the functions directly in some display code in an example), and at least `devicePhys` gives the trunk of the path for the physical device relative to the mapping in `/dev/input/by-path`.  It still is not the name with which the descriptor was opened (although I am skeptical that will ever be reliable), but it would make for better logs.  For example, my applications using evdev all rely on the `by-path` mapping, so that I can label the ports for their purpose and less trained staff can still swap out card readers.

As I said above though I am interested in other use cases and justifications.  I am just trying to get my thoughts out here now :)

Reply:
Well, since I went on the rant about how we can't rely on getting a filepath from the system I went on to get a `FilePath` from the system :rofl: 

So, I do think there is some way to do this with `fnctl`, but honestly just using the standard (and recommened) `/proc/<pid>/fd/<n>` interface provided by linux and calling `readlink` (`readSymbolicLink` in Haskell) seems to work quite well in my tests.  It doesn't return the path-by-id that I used to open the FD, but the event id is correct.  Given that all the hardcore Unix/Linux C people (at least in the usual haunts like `linuxquestions` and the overflows) all seem to say just use that method, but don't rely on there being only one path (as we definitely know working with the linked stuff in `/dev/input/)`, I think this is a better way to go than adding more C to the project expecially since I can't find the documentation for the method I suggested using `fnctl`.

So, hopefully this makes sense.  Let me know if there are any questions, concerns, or insights from your end.  I am optimistic, however, that this will suffice.  I also think it makes sense to continue including this field and to get the path as an upfront cost especially, because there is no guarantee the FD will still be valid after an error and I think error reporting/handling is one of the most important use cases for getting the path associated with the descriptor.

Reply:
> just using the standard (and recommened) /proc/<pid>/fd/<n> interface provided by linux and calling readlink (readSymbolicLink in Haskell) seems to work quite well in my tests

Yep, it's working here, very happy with that.